File: javascript/mnemonic.js
#!/usr/bin/env node

// This is a script for a Mnemonic Number Encoding system
// The MIT License (MIT) - https://opensource.org/licenses/MIT

// Import words list JSON and check whether script is run as a module or standalone
const wordslist = require("../words-list.json");
const isModule = require.main !== module;

// Map each word in the words list to its respective index
const wordsindex = {};
wordslist.map((word, index) => {
  wordsindex[word] = index;
});

// Count the total number of words in the words list
const wordsCount = BigInt(wordslist.length);

// Function to encode a number into a list of words
const encode = (number) => {
  // Check if input is a number
  if (!Number.isInteger(+number)) {
    if (!isModule) console.log("Invalid number");
    return null;
  }

  const words = [];
  let number_to_encode = BigInt(number);

  // Convert the number to a list of words
  while (number_to_encode) {
    words.push(wordslist[number_to_encode % wordsCount]);
    number_to_encode /= wordsCount;
  }

  // Return the encoded words list
  return words.reverse();
};

// Function to decode a list of words back into a number
const decode = (word_list) => {
  // Check if input is a string or array
  if (!Array.isArray(word_list) && typeof word_list !== "string") {
    if (!isModule) console.log("Invalid input");
    return null;
  }

  let number = 0n;
  let l = BigInt(word_list.length - 1);
  for (let i = 0; i < word_list.length; ++i) {
    const word = word_list[i];

    // Check if word exists in the word index
    if (!(word in wordsindex)) {
      if (!isModule) console.error(`'${word}' not found in word list`);
      return null;
    }

    // Convert the list of words back into a number
    const v = wordsindex[word];
    number += BigInt(v) * wordsCount ** l;
    l -= 1n;
  }

  // Return the decoded number
  return number;
};

// Export the functions for usage in other scripts if this script is used as a module
module.exports = {
  encode,
  decode,
};

// Main function for command line usage
(function main(argv) {
  if (isModule) return;
  if (argv.length < 3) {
    return console.error(
      `Provide a number to encode, or a list of words to decode`
    );
  }

  // Check if the first argument is a number or string
  // And encode or decode accordingly
  let res;
  const args = argv.slice(2);
  if (isNaN(args[0])) {
    res = decode(args);
  } else {
    res = encode(args[0]);
  }

  if (res) {
    console.log(res);
  }
})(process.argv);

File: python/mnemonic.py
#!/usr/bin/env python3

# This is a script for a Mnemonic Number Encoding system
# The MIT License (MIT) - https://opensource.org/licenses/MIT

import re
import sys
import json

# Load word list
wordslist = json.loads(open('../words-list.json').read())
words_count = len(wordslist)
is_not_module = __name__ == '__main__'

# Create word to index mapping
words_index = {word: index for index, word in enumerate(wordslist)}

# check if a number is numeric and return its base
def isnumeric(number):
  number = str(number)
  if number.isnumeric():
    return 10
  
  regex = {
    2: r"^0b[01]+$",
    8: r"^0o[0-7]+$",
    16: r"^0x[0-9A-Fa-f]+$",
    32: r"^0v[0-9A-Z]+$",
    64: r"^0n[0-9A-Za-z+/]+$",
  }

  for base in regex:
    if re.match(regex[base], number):
      return base

  return False

def encode(number):
  """Encode a number into a list of words"""
  base = isnumeric(number)
  if not base:
    if is_not_module:
      print('Invalid number')
    return None

  # Convert the number to a list of words
  words = []
  number_to_encode = int(number, base)
  while number_to_encode > 0:
    words.append(wordslist[number_to_encode % words_count])
    number_to_encode //= words_count

  # Return the encoded words list
  return " ".join(reversed(words))


def decode(word_list):
  """Decode a list of words into a number"""

  # Make sure the word_list argument is a valid array or string
  if not isinstance(word_list, (list, str)):
    if is_not_module:
      print('Invalid input')
    return None

  number = 0
  for i, word in enumerate(reversed(word_list)):
    if word not in words_index:
      print(f"'{word}' not found in word list")
      return None

    v = words_index[word]
    number += v * (words_count ** i)

  return number


# Check if the first argument is a number or string
# And encode or decode accordingly
if __name__ == "__main__":
  if len(sys.argv) < 2:
    print('Provide a number to encode, or a list of words to decode')
    sys.exit(1)

  res = None
  args = sys.argv[1:]
  if args[0].isalpha():
    res = decode(args)
  else:
    res = encode(args[0])
  if res:
    print(res)
